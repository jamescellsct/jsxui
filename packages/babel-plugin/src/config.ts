import * as t from '@babel/types'
import template from '@babel/template'
import get from 'dlv'

export function createConfig(
  activeVisitor: 'figma' | 'ink' | 'native' | 'web'
) {
  const theme = {
    breakpoints: {
      small: '@media (min-width: 600px)',
      medium: '@media (min-width: 960px)',
      large: '@media (min-width: 1200px)',
    },
    colors: {
      brand: '#8D6CEE',
      brandTint: '#7B5AD9',
      foreground: 'black',
      foregroundSecondary: 'gray',
    },
    fontSizes: {
      small: {
        initial: 10,
        'breakpoints.medium': 12,
      },
      medium: {
        initial: 12,
        'breakpoints.medium': 16,
      },
      large: {
        initial: 20,
        'breakpoints.medium': 28,
      },
      xlarge: {
        initial: 32,
        'breakpoints.medium': 48,
      },
    },
    fontWeights: {
      medium: 400,
      bold: {
        initial: 500,
        'prefers.dark': 600,
      },
    },
    letterSpacings: {
      normal: 0.25,
      tracked: {
        initial: 0.3,
        'prefers.dark': 0.4,
      },
    },
    boxSpacings: {
      xxsmall: '4px',
      xsmall: '8px',
      small: '16px',
      medium: '24px',
      large: '40px',
      xlarge: '60px',
      xxlarge: '80px',
    },
    boxSizes: {
      container: {
        small: '640px',
        medium: '960px',
        large: '1440px',
      },
      button: {},
      input: {},
      icon: {
        small: '20px',
        medium: '32px',
      },
    },
    variants: {
      heading1: {
        initial: {
          size: 'fontSizes.large',
        },
        web: {
          as: 'h1',
        },
        native: {
          as: 'Text',
          accessibilityRole: 'header',
        },
      },
      heading2: {
        initial: {
          size: 'fontSizes.medium',
        },
        web: {
          as: 'h2',
        },
        native: {
          as: 'Text',
          accessibilityRole: 'header',
        },
      },
    },
  }

  const getValue = (value, key) => {
    return typeof value === 'number' ? value : get(theme[key], value, value)
  }

  const textPlatformComponents = {
    figma: {
      as: 'Text',
      source: 'react-figma',
    },
    ink: {
      as: 'Text',
      source: 'ink',
    },
    native: {
      as: 'Text',
      source: 'react-native',
    },
    web: {
      as: 'span',
    },
  }

  const stackPlatformComponents = {
    figma: {
      as: 'View',
      source: 'react-figma',
    },
    ink: {
      as: 'Box',
      source: 'ink',
      defaults: {
        flexDirection: 'column',
      },
    },
    native: {
      as: 'View',
      source: 'react-native',
    },
    web: {
      as: 'div',
      defaults: {
        display: 'flex',
      },
    },
  }

  const components = [
    {
      name: 'Text',
      transforms: {
        size: (value, theme) => ({
          fontSize: theme.fontSizes[value]?.initial || theme.fontSizes[value],
        }),
        weight: (value, theme) => ({
          fontWeight:
            theme.fontWeights[value]?.initial || theme.fontWeights[value],
        }),
        alignment: (value) => ({
          textAlign: value,
        }),
        opacity: (value) => ({
          opacity: value,
        }),
        color: (value) => ({
          color: theme.colors[value],
        }),
      },
      ...textPlatformComponents[activeVisitor],
    },
    {
      name: 'Stack',
      transforms: {
        axis: (value) => ({
          flexDirection: value === 'x' ? 'row' : 'column',
        }),
        width: (value) =>
          value.includes('fr')
            ? { flex: value.slice(0, -2) }
            : getValue(value, 'boxSpacings'),
        spaceX: (value) => ({
          paddingLeft: getValue(value, 'boxSpacings'),
          paddingRight: getValue(value, 'boxSpacings'),
        }),
        spaceXStart: (value) => ({
          paddingLeft: getValue(value, 'boxSpacings'),
        }),
        spaceXEnd: (value) => ({
          paddingRight: getValue(value, 'boxSpacings'),
        }),
        spaceY: (value) => ({
          paddingTop: getValue(value, 'boxSpacings'),
          paddingBottom: getValue(value, 'boxSpacings'),
        }),
        spaceYStart: (value) => ({
          paddingTop: getValue(value, 'boxSpacings'),
        }),
        spaceYEnd: (value) => ({
          paddingBottom: getValue(value, 'boxSpacings'),
        }),
        spaceBetween:
          activeVisitor === 'web'
            ? (value) => ({
                gap: value,
              })
            : {
                // how can we manipulate children? Wrap them with a special component?
                // can we reference other compiler components here?
                // children: value => ({
                //   <Spacer size={value} />
                // })
              },
        background: (value) => ({
          background: theme.colors[value],
        }),
      },
      ...stackPlatformComponents[activeVisitor],
    },
  ]

  const webVisitor = {
    JSXOpeningElement(path) {
      const id = this.getElementId(path)
      const styleProperties = this.styleProperties[id]
      if (styleProperties) {
        path.node.attributes.push(
          t.jsxAttribute(
            t.jsxIdentifier('css'),
            t.jsxExpressionContainer(t.objectExpression(styleProperties))
          )
        )
      }
    },
  }

  const figmaVisitor = {
    JSXOpeningElement(path) {
      const id = this.getElementId(path)
      const styleProperties = this.styleProperties[id]
      if (styleProperties) {
        path.node.attributes.push(
          t.jsxAttribute(
            t.jsxIdentifier('style'),
            t.jsxExpressionContainer(t.objectExpression(styleProperties))
          )
        )
      }
    },
  }

  const inkVisitor = {
    JSXOpeningElement(path) {
      const id = this.getElementId(path)
      const styleAttributes = this.styleAttributes[id]
      if (styleAttributes) {
        path.node.attributes = path.node.attributes.concat(styleAttributes)
      }
    },
  }

  const buildStylesheet = template(`
const styles = StyleSheet.create(STYLES)
`)

  const nativeVisitor = {
    Program(path) {
      path.pushContainer(
        'body',
        buildStylesheet({
          STYLES: t.objectExpression(
            Object.entries(this.styleProperties).map(([id, styleProperties]) =>
              t.objectProperty(
                t.identifier(id),
                t.objectExpression(styleProperties)
              )
            )
          ),
        })
      )
    },
    ImportDeclaration(path) {
      if (path.node.source.value === 'react-native') {
        path.node.specifiers.push(
          t.importSpecifier(
            t.identifier('StyleSheet'),
            t.identifier('StyleSheet')
          )
        )
      }
    },
    JSXOpeningElement(path) {
      const id = this.getElementId(path)
      path.node.attributes.push(
        t.jsxAttribute(
          t.jsxIdentifier('style'),
          t.jsxExpressionContainer(
            t.memberExpression(t.identifier('styles'), t.identifier(id))
          )
        )
      )
    },
  }

  const visitors = {
    figma: figmaVisitor,
    ink: inkVisitor,
    native: nativeVisitor,
    web: webVisitor,
  }

  return {
    components,
    theme,
    visitor: visitors[activeVisitor],
  }
}
