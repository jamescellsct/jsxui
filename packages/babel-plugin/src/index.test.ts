import pluginTester from 'babel-plugin-tester'

import plugin from './index'
import { createConfig } from './config'

pluginTester({
  plugin,
  pluginName: '@jsxui/babel-plugin',
  pluginOptions: createConfig('web'),
  filename: __filename,
  snapshot: true,
  tests: [
    { fixture: '__fixtures__/site.js' },
    // { fixture: '__fixtures__/visibility.js' },
    // { fixture: '__fixtures__/ink.js' },
    // { fixture: '__fixtures__/multiple-props.js' },
    // { fixture: '__fixtures__/empty.js' },
    // { fixture: '__fixtures__/react-figma.js' },
    // { fixture: '__fixtures__/variants.js' },
    // { fixture: '__fixtures__/props.js' },
    // { fixture: '__fixtures__/simple.js' },
    // { fixture: '__fixtures__/variable.js' },
    // { fixture: '__fixtures__/graphic.js' },
  ],
})
