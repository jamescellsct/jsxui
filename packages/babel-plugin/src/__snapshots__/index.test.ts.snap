// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@jsxui/babel-plugin 1. @jsxui/babel-plugin: 1. @jsxui/babel-plugin 1`] = `

import React from 'react'
import { Text, Spacer, Stack } from '@jsxui/react'

import { Github, Logo, Shapes, Wave } from '../assets'

export default function Index() {
  return (
    <Stack height="minmax(100vh, auto)">
      <div style={{ display: 'grid' }}>
        <Stack style={{ gridArea: '1 / 1' }}>
          <Shapes
            visible={{ initial: false, 'breakpoints.large': true }}
            width="100%"
          />
        </Stack>
        <Stack
          axis="x"
          spaceX="minmax(16px, 1fr)"
          style={{ gridArea: '1 / 1' }}
        >
          <Stack
            width="minmax(auto, 960px)"
            spaceX="1fr"
            spaceYStart="16px"
            spaceYEnd="minmax(16px, 1fr)"
          >
            <Stack
              axis="x"
              width="1fr"
              spaceY="1fr"
              spaceBefore={0}
              spaceAfter={0}
            >
              <Logo width="auto" height="32px" />
              <Spacer />
              <Stack as="a" href="https://github.com/souporserious/jsxui">
                <Github />
              </Stack>
            </Stack>
            <Spacer size="80px" />
            <Text
              width="1fr"
              size="xlarge"
              weight="900"
              lineSpacing={24}
              alignment="center"
              spaceBefore={{
                initial: 0,
                screenMedium: 60,
              }}
              spaceAfter={{
                initial: 0,
                screenMedium: 60,
              }}
            >
              Simple UI Primitives to Help Launch Your next&nbsp;Idea
            </Text>
            <Spacer
              size={{
                initial: '32px',
                screenMedium: '48px',
              }}
            />
          </Stack>
        </Stack>
      </div>
      <Stack
        axis="x"
        width="minmax(100%, 1440px)"
        spaceX={{ 'breakpoints.large': '1fr' }}
        spaceBetween={{ 'breakpoints.large': 'minmax(24px, 0.5fr)' }}
        spaceYStart="40px"
        spaceYEnd="80px"
        background={{ xray: 'url(#diagonalHatch)' }}
        // TODO: add X/YWrap components
        style={{ flexWrap: 'wrap' }}
      >
        {['Elements', 'Variants', 'Overrides', 'Editor'].map((name) => (
          <Stack
            key={name}
            width={{
              initial: '50%',
              'breakpoints.large': '1fr',
            }}
            spaceX="1fr"
            spaceBetween="24px"
          >
            <Text size="large" weight={600} alignment="center" opacity={0.75}>
              {name}
            </Text>
            <Text>Coming Soon</Text>
          </Stack>
        ))}
      </Stack>
      <Wave width="1fr" height="auto" />
      <Spacer size="minmax(80px, 1fr)" background="#7B5AD9" />
    </Stack>
  )
}

      ↓ ↓ ↓ ↓ ↓ ↓

var _jsxuiFileName =
  '/Users/travisarnold/Code/jsxui/packages/babel-plugin/src/__fixtures__/site.js'
import React from 'react'
import { Text, Spacer, Stack } from '@jsxui/react'
import { Github, Logo, Shapes, Wave } from '../assets'
export default function Index() {
  return (
    <Stack
      height="minmax(100vh, auto)"
      __jsxuiSource={{
        fileName: _jsxuiFileName,
        lineNumber: 8,
        columnNumber: 5,
      }}
    >
      <div
        style={{
          display: 'grid',
        }}
        __jsxuiSource={{
          fileName: _jsxuiFileName,
          lineNumber: 9,
          columnNumber: 7,
        }}
      >
        <Stack
          style={{
            gridArea: '1 / 1',
          }}
          __jsxuiSource={{
            fileName: _jsxuiFileName,
            lineNumber: 10,
            columnNumber: 9,
          }}
        >
          <Shapes
            visible={{
              initial: false,
              'breakpoints.large': true,
            }}
            width="100%"
            __jsxuiSource={{
              fileName: _jsxuiFileName,
              lineNumber: 11,
              columnNumber: 11,
            }}
          />
        </Stack>
        <Stack
          axis="x"
          spaceX="minmax(16px, 1fr)"
          style={{
            gridArea: '1 / 1',
          }}
          __jsxuiSource={{
            fileName: _jsxuiFileName,
            lineNumber: 16,
            columnNumber: 9,
          }}
        >
          <Stack
            width="minmax(auto, 960px)"
            spaceX="1fr"
            spaceYStart="16px"
            spaceYEnd="minmax(16px, 1fr)"
            __jsxuiSource={{
              fileName: _jsxuiFileName,
              lineNumber: 21,
              columnNumber: 11,
            }}
          >
            <Stack
              axis="x"
              width="1fr"
              spaceY="1fr"
              spaceBefore={0}
              spaceAfter={0}
              __jsxuiSource={{
                fileName: _jsxuiFileName,
                lineNumber: 27,
                columnNumber: 13,
              }}
            >
              <Logo
                width="auto"
                height="32px"
                __jsxuiSource={{
                  fileName: _jsxuiFileName,
                  lineNumber: 34,
                  columnNumber: 15,
                }}
              />
              <Spacer
                __jsxuiSource={{
                  fileName: _jsxuiFileName,
                  lineNumber: 35,
                  columnNumber: 15,
                }}
              />
              <Stack
                as="a"
                href="https://github.com/souporserious/jsxui"
                __jsxuiSource={{
                  fileName: _jsxuiFileName,
                  lineNumber: 36,
                  columnNumber: 15,
                }}
              >
                <Github
                  __jsxuiSource={{
                    fileName: _jsxuiFileName,
                    lineNumber: 37,
                    columnNumber: 17,
                  }}
                />
              </Stack>
            </Stack>
            <Spacer
              size="80px"
              __jsxuiSource={{
                fileName: _jsxuiFileName,
                lineNumber: 40,
                columnNumber: 13,
              }}
            />
            <Text
              width="1fr"
              size="xlarge"
              weight="900"
              lineSpacing={24}
              alignment="center"
              spaceBefore={{
                initial: 0,
                screenMedium: 60,
              }}
              spaceAfter={{
                initial: 0,
                screenMedium: 60,
              }}
              __jsxuiSource={{
                fileName: _jsxuiFileName,
                lineNumber: 41,
                columnNumber: 13,
              }}
            >
              Simple UI Primitives to Help Launch Your next&nbsp;Idea
            </Text>
            <Spacer
              size={{
                initial: '32px',
                screenMedium: '48px',
              }}
              __jsxuiSource={{
                fileName: _jsxuiFileName,
                lineNumber: 58,
                columnNumber: 13,
              }}
            />
          </Stack>
        </Stack>
      </div>
      <Stack
        axis="x"
        width="minmax(100%, 1440px)"
        spaceX={{
          'breakpoints.large': '1fr',
        }}
        spaceBetween={{
          'breakpoints.large': 'minmax(24px, 0.5fr)',
        }}
        spaceYStart="40px"
        spaceYEnd="80px"
        background={{
          xray: 'url(#diagonalHatch)',
        }} // TODO: add X/YWrap components
        style={{
          flexWrap: 'wrap',
        }}
        __jsxuiSource={{
          fileName: _jsxuiFileName,
          lineNumber: 67,
          columnNumber: 7,
        }}
      >
        {['Elements', 'Variants', 'Overrides', 'Editor'].map((name) => (
          <Stack
            key={name}
            width={{
              initial: '50%',
              'breakpoints.large': '1fr',
            }}
            spaceX="1fr"
            spaceBetween="24px"
            __jsxuiSource={{
              fileName: _jsxuiFileName,
              lineNumber: 79,
              columnNumber: 11,
            }}
          >
            <Text
              size="large"
              weight={600}
              alignment="center"
              opacity={0.75}
              __jsxuiSource={{
                fileName: _jsxuiFileName,
                lineNumber: 88,
                columnNumber: 13,
              }}
            >
              {name}
            </Text>
            <Text
              __jsxuiSource={{
                fileName: _jsxuiFileName,
                lineNumber: 91,
                columnNumber: 13,
              }}
            >
              Coming Soon
            </Text>
          </Stack>
        ))}
      </Stack>
      <Wave
        width="1fr"
        height="auto"
        __jsxuiSource={{
          fileName: _jsxuiFileName,
          lineNumber: 95,
          columnNumber: 7,
        }}
      />
      <Spacer
        size="minmax(80px, 1fr)"
        background="#7B5AD9"
        __jsxuiSource={{
          fileName: _jsxuiFileName,
          lineNumber: 96,
          columnNumber: 7,
        }}
      />
    </Stack>
  )
}


`;
